exchange/bitswap/bitswap.go

```diff
@@ -385,7 +404,15 @@ func (bs *Bitswap) wantNewBlocks(ctx context.Context, bkeys []u.Key) {
 			}
 		}(p)
 	}
-	wg.Wait()
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-ctx.Done():
+	}
```

Comment:
if bs.send hangs, we exit leaking a hung goroutine. 

the downside of this approach is that it  becomes easy to hide hanging things, and thus harder to find problems-- nodes would just slowly deteriorate. 

why is `bs.send` hanging? is there a way to speed that up and respect the context there, instead of bypassing it here?

---

For others following this along-- the codebase is full of examples of both things: places where we return asap ad let the slow-to-finish goroutine finish on its own, and places where we explicitly wait for it to be done, so that we know we've fully finished. there's tradeoffs and cases where we _must_ do one or the other.